File: Patrones de Comportamiento/flyweight.java
import java.util.HashMap;
import java.util.Map;

class TipoVehiculo {
    private String nombre;
    private String color;
    private String sonido;

    public TipoVehiculo(String nombre, String color, String sonido) {
        this.nombre = nombre;
        this.color = color;
        this.sonido = sonido;
    }

    public void mostrar(int x, int y) {
        System.out.println("Mostrando vehículo " + nombre + " en [" + x + ", " + y + "] con color " + color + " y sonido " + sonido);
    }
}

class FabricaVehiculos {
    private static final Map<String, TipoVehiculo> tipos = new HashMap<>();

    public static TipoVehiculo obtenerTipo(String nombre, String color, String sonido) {
        if (!tipos.containsKey(nombre)) {
            System.out.println("Creando nuevo tipo de vehículo: " + nombre);
            tipos.put(nombre, new TipoVehiculo(nombre, color, sonido));
        }
        return tipos.get(nombre);
    }
}

class Vehiculo {
    private int x, y;
    private TipoVehiculo tipo;

    public Vehiculo(int x, int y, TipoVehiculo tipo) {
        this.x = x;
        this.y = y;
        this.tipo = tipo;
    }

    public void dibujar() {
        tipo.mostrar(x, y);
    }
} 

public class flyweight {
    public static void main(String[] args) {
        TipoVehiculo tipoCoche = FabricaVehiculos.obtenerTipo("Coche", "Rojo", "sonido_coche.mp3");
        TipoVehiculo tipoMoto = FabricaVehiculos.obtenerTipo("Moto", "Negra", "sonido_moto.mp3");

        Vehiculo[] vehiculos = {
            new Vehiculo(10, 20, tipoCoche),
            new Vehiculo(50, 20, tipoCoche),
            new Vehiculo(-10, 20, tipoMoto),
            new Vehiculo(0, 0, tipoMoto)
        };

        for (Vehiculo vehiculo : vehiculos) {
            vehiculo.dibujar();
        }
    }
}

File: Patrones de Comportamiento/flyweight.py

class TipoEnemigo:
    def __init__(self, nombre, textura, sonido):
        self.nombre = nombre
        self.textura = textura
        self.sonido = sonido

    def mostrar(self, x, y):
        print(f"Mostrando {self.nombre} en [{x}, {y}] con texto {self.textura} y sonido {self.sonido}")

class FabricaEnemigos:
    _tipos = {}

    @staticmethod
    def obtener_tipo(nombre, textura, sonido):
        if nombre not in FabricaEnemigos._tipos:
            print(f"Creando nuevo tipo de enemigos: {nombre}")
            FabricaEnemigos._tipos[nombre] = TipoEnemigo(nombre, textura, sonido)
        return FabricaEnemigos._tipos[nombre]

class Enemigo:
    def __init__(self, x, y, tipo):
        self.tipo = tipo
        self.x = x
        self.y = y
        
    def dibujar(self):
        self.tipo.mostrar(self.x, self.y)

tipo_orco = FabricaEnemigos.obtener_tipo("Orco", "orco.png", "sonido_orco.mp3")
tipo_elfo = FabricaEnemigos.obtener_tipo("Elfo", "elfo.png", "sonido_elfo.mp3")

enemigos = [
    Enemigo(10, 20, tipo_orco),
    Enemigo(50, 20, tipo_orco),
    Enemigo(-10, 20, tipo_elfo),
    Enemigo(0, 0, tipo_elfo),
]


for enemigo in enemigos:
    enemigo.dibujar()